spacy/pipeline/span_ruler.py

```diff
@@ -0,0 +1,345 @@
+    name: str,
+    spans_key: str,
+    phrase_matcher_attr: Optional[Union[int, str]],
+    validate: bool,
+    overwrite: bool,
+    scorer: Optional[Callable],
+):
+    return SpanRuler(
+        nlp,
+        name,
+        spans_key=spans_key,
+        phrase_matcher_attr=phrase_matcher_attr,
+        validate=validate,
+        overwrite=overwrite,
+        scorer=scorer,
+    )
+
+
+class SpanRuler(Pipe):
+    """The SpanRuler lets you add spans to the `Doc.spans` using token-based
+    rules or exact phrase matches.
+
+    DOCS: https://spacy.io/api/spanruler
+    USAGE: https://spacy.io/usage/rule-based-matching#spanruler
+    """
+
+    def __init__(
+        self,
+        nlp: Language,
+        name: str = "span_ruler",
+        *,
+        spans_key: str = "ruler",
+        phrase_matcher_attr: Optional[Union[int, str]] = None,
+        validate: bool = False,
+        overwrite: bool = False,
+        scorer: Optional[Callable] = spancat_score,
+        patterns: Optional[List[PatternType]] = None,
+    ) -> None:
+        """Initialize the span ruler. If patterns are supplied here, they
+        need to be a list of dictionaries with a `"label"` and `"pattern"`
+        key. A pattern can either be a token pattern (list) or a phrase pattern
+        (string). For example: `{'label': 'ORG', 'pattern': 'Apple'}`.
+
+        nlp (Language): The shared nlp object to pass the vocab to the matchers
+            and process phrase patterns.
+        name (str): Instance name of the current pipeline component. Typically
+            passed in automatically from the factory when the component is
+            added. Used to disable the current span ruler while creating
+            phrase patterns with the nlp object.
+        spans_key (str): The spans key to save the spans under. Defaults to
+            "ruler".
+        phrase_matcher_attr (int / str): Token attribute to match on, passed
+            to the internal PhraseMatcher as `attr`
+        validate (bool): Whether patterns should be validated, passed to
+            Matcher and PhraseMatcher as `validate`
+        patterns (iterable): Optional patterns to load in.
+        overwrite (bool): Whether to reset any existing spans under this spans
+            key.
+        scorer (Optional[Callable]): The scoring method. Defaults to
+            spacy.pipeline.spancat.spancat_score.
+
+        DOCS: https://spacy.io/api/spanruler#init
+        """
+        self.nlp = nlp
+        self.name = name
+        self.cfg = {
+            "spans_key": spans_key,
+            "overwrite": overwrite,
+            "phrase_matcher_attr": phrase_matcher_attr,
+            "validate": validate,
+        }
+        self.clear()
+        if patterns:
+            self.add_patterns(patterns)
+        self.scorer = scorer
+
+    def __len__(self) -> int:
+        """The number of all labels added to the span ruler."""
+        return len(set([p["label"] for p in self._patterns]))
```

Comment:
The `entity_ruler` defines `__len__` as the number of patterns, so this definition might be unexpected?

Reply:
It's unclear what else you might count, though. I am open to suggestions.

Reply:
Ah, I think was originally trying not to track the added patterns to not have things duplicated in the component and the matchers, but added it back at some point to make a few other functions easier. We could reconsider that, I guess, although I think not serializing the patterns in their original input format has some drawbacks for users who might want to edit the patterns directly.